### PR TITLE
o added even driven streaming with fd-mux

### PR DESCRIPTION
**Pull request by lromor**

### CODE REVIEW

File: src/gcode-streamer_test.cc [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  int NewData(const char *data) {
+    // Feed data in the pipe
+    return write(fd_[1], data, strlen(data));
+  }
+
+  void ClientDisconnect() { close(fd_[1]); }
+
+private:
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: Tip: I'd add some constants (e.g. an enum) to contain the numbers of the ends of the pipe
```
enum { RECEIVING_FD, SENDING_FD };
```

Then above, it is much easier to read what we're using via the symbolic names e.g.
```
  int SendData(const char *data) {
      return write(fd_[SENDING_FD], data, strlen(data));
  }
```

File: src/gcode-streamer_test.cc [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  {
+    EXPECT_CALL(tester, gcode_start(_)).Times(0);
+    EXPECT_CALL(tester, coordinated_move(_, _)).Times(0);
+    EXPECT_CALL(tester, gcode_finished(_)).Times(1);
+  }
+  // Connect
+  tester.OpenStream();
+  // Write incomplete move command
+  tester.ReceiveString("G1X200F1000");
+  // Close the stream
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: Don't comment the obvious. A commen that is saying exactly the same as the method we call is just wasting space (also in other places in this file, e.g. the `// Connect` above.

File: src/gcode-streamer_test.cc [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  StreamTester tester;
+  {
+    EXPECT_CALL(tester, gcode_start(_)).Times(0);
+    EXPECT_CALL(tester, coordinated_move(_, _)).Times(0);
+    EXPECT_CALL(tester, gcode_finished(_)).Times(1);
+  }
+  // Connect
+  tester.OpenStream();
+  // Write incomplete move command
+  tester.ReceiveString("G1X200F1000");
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: To save vertical space, maybe just put the short comments just behind the code:
```
 tester.ReceiveString("G1X200F1000"); // write incomplete commmand.
```

File: src/gcode-streamer_test.cc [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    stream_mock_ = new MockStream();
+    return streamer_->ConnectStream(stream_mock_->ClientConnect(), NULL);
+  }
+
+  void CloseStream() {
+    stream_mock_->ClientDisconnect();
+    delete stream_mock_;
+    stream_mock_ = NULL;
+  }
+  void ReceiveString(const char *line) {
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: This `ReceiveString()` is very hard to understand, as it actually describes the opposite what it does.
It should probably be named `SendString()`.

File: src/gcode-streamer_test.cc [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  virtual void wait_temperature() {}
+  virtual void dwell(float time_ms) {}
+  virtual void motors_enable(bool enable) {}
+  virtual bool rapid_move(float feed_mm_p_sec,
+                          const AxesRegister &axes) { return true;}
+  virtual const char *unprocessed(char letter, float value,
+                                  const char *rest_of_line) { return NULL; }
+
+private:
+
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: Don't leave unnecessary newlines after `private:`.
There should always be a newline before `protected:` or `private:`, but not after.

File: src/machine-control.cc [hunk 6]
```diff
@@ -192,19 +196,27 @@ static int run_server(int listen_socket,
...
+                              [listen_socket,machine,streamer]() {
     struct sockaddr_in client;
     socklen_t socklen = sizeof(client);
     struct sigaction sa = {};
     sa.sa_handler = eatsignal;
     sa.sa_flags = SA_RESETHAND | SA_NODEFER;  // oneshot, no restart
     sigaction(SIGINT, &sa, NULL);
     sigaction(SIGTERM, &sa, NULL);
-    int connection = accept(listen_socket, (struct sockaddr*) &client, &socklen);
+    int connection = accept4(listen_socket, (struct sockaddr*) &client,
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/machine-control.cc]: (solved)

File: src/machine-control.cc [hunk 7]
```diff
@@ -214,19 +226,14 @@ static int run_server(int listen_socket,
...
+    streamer->ConnectStream(connection, msg_stream);
+    return true;
+  });
 
-  close(listen_socket);
-  Log_error("Error gcode_machine_control_from_stream() == %d. Exiting\n",
-            process_result);
-
-  return process_result;
+  return event_server->Loop();
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/machine-control.cc]: Ok, I kept it inside the run_server function just to be consistent with the send_file_to_machine

File: src/gcode-streamer_test.cc (hunk 5 above)
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc (hunk 4 above)
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc (hunk 3 above)
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc (hunk 2 above)
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+#include "gcode-streamer.h"
+
+class MockStream {
+public:
+
+  MockStream() : fd_{-1, -1} {}
+  ~MockStream() { ClientDisconnect(); }
+
+  int ClientConnect() {
+    ClientDisconnect();
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+#include "common/fd-mux.h"
+#include "common/logging.h"
+#include "gcode-parser/gcode-parser.h"
+#include "gcode-streamer.h"
+
+class MockStream {
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/common/linebuf-reader_test.cc [hunk 10]
```diff
@@ -0,0 +1,125 @@
+/* -*- mode: c++; c-basic-offset: 2; indent-tabs-mode: nil; -*-
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/linebuf-reader_test.cc]: (solved) but with:
 * (c) 2016 Henner Zeller <h.zeller@acm.org>

File: src/common/fd-mux.cc [hunk 11]
```diff
@@ -0,0 +1,162 @@
+// -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil; -*-
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.cc]: (solved)

File: src/common/fd-mux.h [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+#include <map>
+#include <functional>
+#include <vector>
+
+// This needs a better name.
+class FDMultiplexer {
+public:
+
+  // Handler for connections. Returns '1' if we want to continue reading
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/common/fd-mux.h [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+  // Schedules the delete of an fd,callback pair for the next loop cycle.
+  void ScheduleDelete(int fd);
+
+  // Run the main loop. Blocks while there is still a filedescriptor
+  // registered (return 0) or until a signal is triggered (return 2).
+  int Loop();
+
+protected:
+  bool Cycle(unsigned timeout_ms = 50);
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/common/fd-mux.cc [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+int FDMultiplexer::Loop() {
+
+  arm_signal_handler();
+  while (Cycle(10)) {}
+  disarm_signal_handler();
+
+  if (caught_signal)
+    return 2;
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.cc]: (solved)

File: src/common/fd-mux.h [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+  bool RunOnReadable(int fd, const Handler &handler);
+  bool RunOnWritable(int fd, const Handler &handler);
+  void RunOnTimeout(const Handler &handler);
+
+  // Schedules the delete of an fd,callback pair for the next loop cycle.
+  void ScheduleDelete(int fd);
+
+  // Run the main loop. Blocks while there is still a filedescriptor
+  // registered (return 0) or until a signal is triggered (return 2).
+  int Loop();
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: ahah, I was leaving space for other types of errors indeed =P no offense to 1.

File: src/gcode-streamer.cc [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  if (reader_.Update(connection_fd_) == 0) {
+    // Let's test if the connection is still active
+    char buffer;
+    if (recv(connection_fd_, &buffer,
+             sizeof(buffer), MSG_PEEK | MSG_DONTWAIT) == 0) {
+      Log_info("Client disconnected.");
+    } else {
+      Log_info("Reached EOF.");
+    }
+
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer.cc]: (solved)

File: src/gcode-streamer_test.cc [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+using namespace ::testing;
+
+// The new line is the "send" character. If we disconnect and the linebuffer
+// has not completed the last line, we should drop it
+// NOTE: Broken
+TEST(Streaming, no_newline_no_parsing) {
+  StreamTester tester;
+  {
+    EXPECT_CALL(tester, gcode_start(_)).Times(0);
+    EXPECT_CALL(tester, coordinated_move(_, _)).Times(0);
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+class MockStream {
+public:
+
+  MockStream() : fd_{-1, -1} {
+    if (pipe(fd_) < 0) {
+      perror("pipe()");
+      exit(1);
+    }
+  }
+  ~MockStream() { CloseSender(); }
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer_test.cc [hunk 19]
```diff
@@ -0,0 +1,190 @@
+/* -*- mode: c++; c-basic-offset: 2; indent-tabs-mode: nil; -*-
+ * Test for pru motion queue.
+ *
+ * We simulate the pru hardware and test if the absolute position is correctly
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: This header comment looks like it has been copied from somewhere else :)

File: src/gcode-streamer_test.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (hzeller) [reply src/gcode-streamer_test.cc]: Or, we just don't do anything in the destructor, as, if everything goes right, the receiver should have closed its side of the filedescriptor.

File: src/common/fd-mux.h [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+  // '-1' Can be used to exit the multiplexer with an error, for example
+  // if one of our tasklets has an error.
+  typedef std::function<bool()> Handler;
+
+  // These can only be set before Loop() is called or from a
+  // running handler itself.
+  // Returns false if that filedescriptor is already registered.
+  bool RunOnReadable(int fd, const Handler &handler);
+  bool RunOnWritable(int fd, const Handler &handler);
+  void RunOnTimeout(const Handler &handler);
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/common/fd-mux.h [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+  bool Cycle(unsigned timeout_ms = 50);
+
+private:
+
+  std::map<int, Handler> r_handlers_;
+  std::map<int, Handler> w_handlers_;
+  std::vector<Handler> t_handlers_;
+  std::vector<int> to_delete_r_;
+  std::vector<int> to_delete_w_;
+
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/gcode-streamer_test.cc (hunk 19 above)
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved) *facepalm*

File: src/gcode-streamer_test.cc [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+
+  MockStream() : fd_{-1, -1} {
+    if (pipe(fd_) < 0) {
+      perror("pipe()");
+      exit(1);
+    }
+  }
+  ~MockStream() { CloseSender(); }
+
+  enum { RECEIVING_FD, SENDING_FD };
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/machine-control.cc [hunk 23]
```diff
@@ -494,17 +515,24 @@ int main(int argc, char *argv[]) {
...
                                filename, file_loop_count);
   } else {
-    ret = run_server(listen_socket, machine_control, parser,
+    ret = run_server(listen_socket, &event_server, machine_control, streamer,
                      bind_addr, listen_port);
   }
 
+  if (listen_socket >= 0) {
+    close(listen_socket);
+    Log_error("Exiting server");
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/machine-control.cc]: (solved)

File: src/machine-control.cc [hunk 24]
```diff
@@ -494,17 +515,24 @@ int main(int argc, char *argv[]) {
...
-    ret = send_file_to_machine(machine_control, parser,
+    ret = send_file_to_machine(&event_server, machine_control, streamer,
                                filename, file_loop_count);
   } else {
-    ret = run_server(listen_socket, machine_control, parser,
+    ret = run_server(listen_socket, &event_server, machine_control, streamer,
                      bind_addr, listen_port);
   }
 
+  if (listen_socket >= 0) {
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/machine-control.cc]: (solved)

File: src/gcode-streamer_test.cc [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+    event_server_.Cycle(0);
+  }
+
+  MOCK_METHOD1(gcode_start, void(GCodeParser *parser));
+  MOCK_METHOD1(gcode_finished, void(bool end_of_stream));
+  MOCK_METHOD1(input_idle, void(bool is_first));
+  MOCK_METHOD2(coordinated_move,
+               bool(float feed_mm_p_sec, const AxesRegister &absolute_pos));
+
+  virtual void go_home(AxisBitmap_t axis_bitmap) {}
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/gcode-streamer.cc [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// New data to be fed into the linebuffer
+bool GCodeStreamer::ReadData() {
+
+  // Update buffer
+  if (reader_.Update(connection_fd_) == 0) {
+    // Let's test if the connection is still active
+    char buffer;
+    if (recv(connection_fd_, &buffer,
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer.cc]: (solved)

File: src/gcode-streamer.h [hunk 27]
```diff
@@ -0,0 +1,37 @@
+
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer.h]: (solved)

File: src/gcode-streamer_test.cc [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+};
+
+using namespace ::testing;
+
+// The new line is the "send" character. If we disconnect and the linebuffer
+// has not completed the last line, we should drop it
+// NOTE: Broken
+TEST(Streaming, no_newline_no_parsing) {
+  StreamTester tester;
+  {
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer_test.cc]: (solved)

File: src/common/fd-mux.h [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+  // Schedules the delete of an fd,callback pair for the next loop cycle.
+  void ScheduleDelete(int fd);
+
+  // Run the main loop. Blocks while there is still a filedescriptor
+  // registered (return 0) or until a signal is triggered (return 1).
+  int Loop();
+
+protected:
+  // Useful for testing to control the loop.
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/common/fd-mux.h [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+  bool SingleCycle(unsigned timeout_ms);
+
+private:
+  const unsigned timeout_ms_;
+  std::map<int, Handler> r_handlers_;
+  std::map<int, Handler> w_handlers_;
+  std::vector<Handler> t_handlers_;
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/common/linebuf-reader.h [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+#ifndef _BEAGLEG_LINEBUF_READER_H
+#define _BEAGLEG_LINEBUF_READER_H
+
+#include <stddef.h>
+#include <functional>
+#include <unistd.h>
+
+// A reader to be used in conjunction with file descriptor event management
+// such as select() or poll(). Whenever a file descriptor is readable, it can
+// be used to udpate this LinebufReader. Subsequent calls to
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/linebuf-reader.h]: (solved)

File: src/common/fd-mux.h [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+#include <map>
+#include <functional>
+#include <vector>
+
+// This needs a better name.
+class FDMultiplexer {
+public:
+  FDMultiplexer(unsigned timeout_ms = 50) : timeout_ms_(timeout_ms) {}
+
+  // Handler for connections. Returns true if we want to continue reading
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/common/fd-mux.h]: (solved)

File: src/common/linebuf-reader.h [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+  // The tpical way this will be called: with a file descriptor that has some
+  // bytes ready.
+  int Update(int fd) {
+    return Update([fd](char *buf, size_t len) {
+            return read(fd, buf, len);
+           });
+  }
+
+  // Return a current line if it is available. The line is a nul terminated
+  // c-string.
```
Comment (hzeller) [inline src/common/linebuf-reader.h]: Here, we need to document if or if not it contains the newline character.

File: src/common/fd-mux.h [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+  // we encountered a timeout and the idle-Handler has been called.
+  // This is broken out to make it simple to test steps in unit tests.
+  bool SingleCycle(unsigned timeout_ms);
+
+private:
+  const unsigned timeout_ms_;
+  std::map<int, Handler> read_handlers_;
+  std::map<int, Handler> write_handlers_;
+  std::vector<Handler> idle_handlers_;
+  std::vector<int> to_delete_read_;
```
Comment (hzeller) [inline src/common/fd-mux.h]: These to_delete vectors look like they can be local variables to the SingleCycle() function and don't need to be vectors here.

File: src/common/fd-mux.cc [hunk 35]
```diff
@@ -0,0 +1,172 @@
...
+  timeout.tv_sec = timeout_ms / 1000;
+  timeout.tv_usec = (timeout_ms % 1000) * 1000;
+
+  int maxfd = -1;
+  FD_ZERO(&read_fds);
+  FD_ZERO(&write_fds);
+
+  // Readers
+  for (const auto &it : read_handlers_) {
+    if (it.first >= maxfd) maxfd = it.first + 1;
```
Comment (hzeller) [inline src/common/fd-mux.cc]: Here, we are mixing two things which might be harder to understand for the reader.
You are doing `it.first + 1`, and the reaser might ask why not just it.fisrt.
This is because select() wants the highest filedescriptor + 1 later. But the reader does not know that yet, and the variable does not reflect that (after all, it is not called `max_fd_plus_one`).
And this also means that we have this **>=** test, which is a bit confusing.

I suggest that we just collect the hightest fds here, simply with
```
   maxfd = std::max(maxfd, it.first);
```

and then later, in the call to select do
```
  select(maxfd + 1, ....)
```

This helps reading the code.

File: src/gcode-streamer.cc [hunk 36]
```diff
@@ -0,0 +1,109 @@
...
+#include "gcode-streamer.h"
+
+
+GCodeStreamer::GCodeStreamer(FDMultiplexer *event_server, GCodeParser *parser,
+                             GCodeParser::EventReceiver *parse_events)
+  : event_server_(event_server), parser_(parser), parse_events_(parse_events),
+    reader_(), is_streaming_(false), is_processing_(false), connection_fd_(-1) {
+  // Let's start the input idle tasklet
+  event_server_->RunOnIdle([this](){
+    return Timeout();
```
[the remark this answers is not in the document]
Comment (lromor) [reply src/gcode-streamer.cc]: (solved)

File: src/common/linebuf-reader.h [hunk 37]
```diff
@@ -81,6 +81,8 @@ class LinebufReader {
   const char *const buffer_end_;
   char *content_start_;
   char *content_end_;
+
+  enum CH_STATE { IN_LINE, CR_SEEN, LF_SEEN } ch_state_;
```
Comment (hzeller) [inline src/common/linebuf-reader.h]: To make the name a little more expressive. Maybe `parse_state_` ?

File: src/common/linebuf-reader.cc [hunk 38]
```diff
@@ -49,11 +50,25 @@ const char* LinebufReader::IncompleteLine() {
...
   for (char *i = content_start_; i < content_end_; ++i) {
-    if (*i == '\n' || *i == '\r') {
-      *i = '\0';
-      const char *line = content_start_;
-      content_start_ = i + 1;
-      return line;
+    switch (ch_state_) {
+      case LF_SEEN: {
+        // Silently ignore newline
+        if (*i == '\n') {
```
Comment (hzeller) [inline src/common/linebuf-reader.cc]: You are using the names opposite of the meaning of the characters here.
CR = Carriage Return = `\r`, and LF = Line Feed = `\n`.

File: src/common/linebuf-reader.cc [hunk 39]
```diff
@@ -49,11 +50,25 @@ const char* LinebufReader::IncompleteLine() {
...
+          ch_state_ = CR_SEEN;
+          content_start_ = i + 1;
+          continue;
+        }
+      }
+      case CR_SEEN: case IN_LINE: {
```
Comment (hzeller) [inline src/common/linebuf-reader.cc]: For readability, I'd put the different cases under each other:
```
case CR_SEEN:  /* fallthrough */
case IN_LINE:  {
   // ...
}
```

Anyway, looking at this, it seems my suggestion of three states was not needed: one of these is redundant. We only really need `IN_LINE` and `CR_SEEN`.

State transitions  @(`{START_STATE} -- {character} --> [{optional action} {NEW_STATE}`)
```
IN_LINE -- \n --> [emitting line] IN_LINE
IN_LINE -- \r --> [emitting line] CR_SEEN
IN_LINE --(any other char) --> IN_LINE
CR_SEEN -- \n --> [skip character] --> IN_LINE
CR_SEEN -- \r --> [emit empty line] --> CR_SEEN
CR_SEEN --(any other char) --> IN_LINE;
```

File: src/common/linebuf-reader.h (hunk 37 above)
Comment (lromor) [reply src/common/linebuf-reader.h]: (solved, see below)

File: src/common/linebuf-reader.cc (hunk 38 above)
Comment (lromor) [reply src/common/linebuf-reader.cc]: (solved, see below)

File: src/common/linebuf-reader.cc (hunk 39 above)
Comment (lromor) [reply src/common/linebuf-reader.cc]: (solved) yes I saw that the only important state is the CR_SEEN so I converted the mini fsm into a single bool. I was going to work on that but you anticipated me :smiley:

File: src/common/linebuf-reader_test.cc [hunk 40]
```diff
@@ -0,0 +1,148 @@
...
+
+#include <string>
+#include <gtest/gtest.h>
+
+// Raw lines that we use as samples.
+constexpr int kSampleLineCount = 3;
+const char *kSampleLines[kSampleLineCount + 1] = {
+  "This is a short line",
+  "This is another, longer line",
+  "",   // Emtpy line
```
Comment (hzeller) [inline src/common/linebuf-reader_test.cc]: Just for extra testing: can we have an additional empty line here (with kSampleLineCount = 4) to see if we can walk the state machine in all cases to detect two separate lines here ?
(and probably an additional `EXPECT_EQ(std::string(kSampleLines[3]), reader.ReadLine());` where we read it.

File: src/gcode-streamer_test.cc [hunk 41]
```diff
@@ -0,0 +1,180 @@
...
+                          const char *rest_of_line) override { return NULL; }
+
+private:
+  class MockFDMultiplexer : public FDMultiplexer {
+  public:
+    friend class StreamTester;
+  };
+
+  MockFDMultiplexer event_server_;
+  std::unique_ptr<GCodeParser> parser_;
```
Comment (hzeller) [inline src/gcode-streamer_test.cc]: `std::unique_ptr<>` requires the memory-header
```
#include <memory>
```

File: src/common/linebuf-reader_test.cc [hunk 42]
```diff
@@ -0,0 +1,148 @@
...
+  }
+  EXPECT_EQ(kSampleLineCount, expected_next_sample);
+}
+
+INSTANTIATE_TEST_CASE_P(PortableEndLineTests,
+                        LinebufReaderTest,
+                        ::testing::Values("\n", "\r", "\r\n"));
+
+// TODO(hzeller): more testing
+//   - what happens if we have \r\n at the end. Ideally, we would like to see
```
Comment (hzeller) [inline src/common/linebuf-reader_test.cc]: This comment can probably be removed now.

**Comment by lromor** [comment]
I have rebased the branch to fast forward

**Comment by hzeller** [comment]
Thanks!
I have now merged this and split into three submits:
 https://github.com/hzeller/beagleg/commit/5277157d4048d3e007c25e9d43377642d82badc8
 https://github.com/hzeller/beagleg/commit/b4ca45e46468b8b7207cf524ee230f3f7a9090a3
 https://github.com/hzeller/beagleg/commit/8b5e1344b3c2704aa7e75f05f37f229e923f2608

I have done some basic test, reading from a file and using `./machine-control -n -p9999 -c sample.config` to simulate a machine and connect to it locally and found that I needed to remove the signal handling stuff in the `run_server()`, as it was interfering with fd-mux was already doing. Great: less code in `run_server()`.
